api/app/GraphQL/Queries/CountPoolCandidatesByPool.php

```diff
@@ -0,0 +1,103 @@
+use Illuminate\Support\Facades\Log;
+use App\Models\Pool;
+
+final class CountPoolCandidatesByPool
+{
+    /**
+     * @param  null  $_
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        $filters = $args["where"];
+
+        // query counts pool candidate rows, so start on that model
+        $queryBuilder = PoolCandidate::query();
+
+        // pool candidate filters go here
+
+        // candidate pool scope
+        if (array_key_exists('pools', $filters)) {
+            // pool candidate filter uses Pool while Applicant filter users IdInput
+            $pools = array_map(function ($id) {
+                return ['id' => $id];
+            }, $filters['pools']);
+            PoolCandidate::filterByPools($queryBuilder, $pools);
+        }
+
+        $queryBuilder->whereHas('user', function (Builder $userQuery) use ($filters) {
+            // user filters go here
+
+            // user status scope
+            User::scopeAvailableForOpportunities($userQuery);
+
+            // has diploma
+            if (array_key_exists('hasDiploma', $filters)) {
+                User::scopeHasDiploma($userQuery, $filters['hasDiploma']);
+            }
+
+            // equity
+            if (array_key_exists('equity', $filters)) {
+                User::filterByEquity($userQuery, $filters['equity']);
+            }
+
+            // languageAbility
+            if (array_key_exists('languageAbility', $filters)) {
+                User::filterByLanguageAbility($userQuery, $filters['languageAbility']);
+            }
+
+            // operationalRequirements
+            if (array_key_exists('operationalRequirements', $filters)) {
+                User::filterByOperationalRequirements($userQuery, $filters['operationalRequirements']);
+            }
+
+            // locationPreferences
+            if (array_key_exists('locationPreferences', $filters)) {
+                User::filterByLocationPreferences($userQuery, $filters['locationPreferences']);
+            }
+
+            // wouldAcceptTemporary
+            if (array_key_exists('wouldAcceptTemporary', $filters)) {
+                User::scopeWouldAcceptTemporary($userQuery, $filters['wouldAcceptTemporary']);
+            }
+
+            // expectedClassifications
+            if (array_key_exists('expectedClassifications', $filters)) {
+                User::scopeClassifications($userQuery, $filters['expectedClassifications']);
+            }
+
+            // skills
+            if (array_key_exists('skills', $filters)) {
+                User::filterBySkills($userQuery, $filters['skills']);
+            }
+        });
+
+        $databaseRows = $queryBuilder
+            ->selectRaw('count(*) as candidate_count, pool_id')
+            ->groupBy('pool_id')
+            ->get();
+
+        // would be nice to do this in the same query as above but alas...
+        $allPools = Pool::whereIn('id', $databaseRows->pluck('pool_id'))->get();
```

Comment:
If I understand what you are getting at, I believe you can do this with the same query by [eager loading](https://laravel.com/docs/9.x/eloquent-relationships#eager-loading) the relationship. I think something like this would work:

```php
return $queryBuilder
  ->with('pool')
  ->select('pool_id', DB::raw('count(*) as candidate_count'))
  ->groupBy('pool_id')
  ->get();
```

You can then tell laravel to change the `candidate_count` to `candidateCount` in the `PoolCandidate` model with:

```php
public function getCandidateCountAttribute()
{
    return $this->attributes['candidate_count'];
}
```

Reply:
I'm not 100% sure this produces the correct result, but it might help get you on the right track to making it a single query hopefully :)

Reply:
Whoah, this works!  Yer a wizard!  :mage: 

It actually still queries the database a second time to pull in the pools but the code is so much cleaner now that the framework does it for me.  I'll take it as a win.  :smile:  Thanks!

[Use query builder to attach pools](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4455/commits/c1ae4ec91616caefe3ad6a84e5441275edcc257f)